### PR TITLE
rpcclient: add detailed error types

### PR DESCRIPTION
**Pull request by yyforyongyu**
This PR adds detailed error types so the caller can use the structured errors to gain finer control, eg, handling an error resulted from an RBF attempt.

### CODE REVIEW

**Review by ziggie1984 — APPROVED** [review]
LGTM, just had a few Nits regarding the comments otherwise straightforward PR.

File: rpcclient/errors.go [hunk 1]
```diff
@@ -15,163 +16,483 @@ var (
...
+// - https://github.com/bitcoin/bitcoin/blob/master/test/functional/rpc_rawtransaction.py#L342
+// - https://github.com/bitcoin/bitcoin/blob/master/test/functional/data/invalid_txs.py
+// - https://github.com/bitcoin/bitcoin/blob/master/test/functional/mempool_accept.py
+// - https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp
+const (
+	// ErrMissingInputsOrSpent is returned when calling
+	// `sendrawtransaction` with missing inputs.
+	ErrMissingInputsOrSpent BitcoindRPCErr = iota
+
+	// ErrMissingInputs is returned when calling `sendrawtransaction` with
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply rpcclient/errors.go:44]: fixed!

File: rpcclient/errors.go [hunk 2]
```diff
@@ -15,163 +16,483 @@ var (
...
+	// RBF is not enabled.
 	//
-	// NOTE: This is not an exhaustive list of errors, but it covers the
-	// usage case of LND.
+	// NOTE: BIP rule 1.
+	ErrMempoolConflict
+
+	// When a transaction adds new unconfirmed inputs.
+	//
+	// NOTE: BIP rule 2.
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply rpcclient/errors.go:69]: fixed!

File: rpcclient/errors.go [hunk 3]
```diff
@@ -15,163 +16,483 @@ var (
...
+	// // allowed.
+	// "unmatched btcd error 4": ErrSameNonWitnessData,
+}
+
+// Define an empty bitcoind errors slice, and initialize it with the errors
+// from the above RPCError during package init.
+var BitcoindErrors []error
+
+// MapRPCErr takes an error returned from calling RPC methods from various
+// chain backend and map it to an defined error here. It uses the `TxErrMap`
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply rpcclient/errors.go:448]: fixed!

File: rpcclient/errors.go [hunk 4]
```diff
@@ -15,163 +16,483 @@ var (
...
+	errSentinel
 )
 
-// MapBtcdErrToRejectReason takes an error returned from
-// `CheckMempoolAcceptance` and maps the error to a bitcoind reject reason.
-func MapBtcdErrToRejectReason(err error) string {
-	// Get the error string and turn it into lower case.
-	btcErr := strings.ToLower(err.Error())
+// Error implements the error interface. It returns the error message defined
+// in `bitcoind`.
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply rpcclient/errors.go:187]: updated the comments! Think we need to find a way to automatically create these errors, maybe parsing the package https://github.com/bitcoin/bitcoin/tree/master/test/functional

File: rpcclient/errors.go [hunk 5]
```diff
@@ -15,163 +16,483 @@ var (
...
 
-		// Some nonstandard transactions - muiltiple OP_RETURNs.
-		"more than one transaction output in a nulldata script": "multi-op-return",
+	// A transaction that is locked by BIP68 sequence logic.
+	ErrNonBIP68Final
 
-		// A timelocked transaction.
-		"transaction is not finalized":               "non-final",
-		"tried to spend coinbase transaction output": "non-final",
+	// Minimally-small transaction(in non-witness bytes) that is allowed.
```
[the remark this answers is not in the document]
Comment (yyforyongyu) [reply rpcclient/errors.go:174]: yeah totally wrong🤦🏻 now fixed

File: rpcclient/errors.go [hunk 6]
```diff
@@ -15,163 +16,513 @@ var (
...
+		strings.ToLower(strippedErrStr),
+		strings.ToLower(strippedMatchStr),
+	)
+}
+
+// Init initializes the bitcoind errors slice with the errors from the
+// `BitcoindRPCErr` defined above.
+func init() {
+	BitcoindErrors = make([]error, 0, errSentinel)
+	for i := uint32(0); i < uint32(errSentinel); i++ {
```
Comment (guggero) [inline rpcclient/errors.go:524]: Can't we just use this loop in `MapRPCErr` above? That way we don't need a package level `init` and also don't have to worry about anyone modifying the global variable.

File: rpcclient/errors.go [hunk 7]
```diff
@@ -15,163 +16,513 @@ var (
...
+	// If not matched, return the original error wrapped.
+	return fmt.Errorf("%w: %v", ErrUndefined, rpcErr)
+}
+
+// matchErrStr takes an error returned from RPC client and matches it against
+// the specified string. If the expected string pattern is found in the error
+// passed, return true. Both the error strings are normalized before matching.
+func matchErrStr(err error, s string) bool {
+	// Replace all dashes found in the error string with spaces.
+	strippedErrStr := strings.Replace(err.Error(), "-", " ", -1)
```
Comment (guggero) [inline rpcclient/errors.go:508]: nit: can use `strings.ReplaceAll()`.

**Review by guggero — APPROVED** [review]
Nice!

File: rpcclient/errors.go (hunk 6 above)
Comment (yyforyongyu) [reply rpcclient/errors.go]: Nice catch! Yeah totally, now switched!

File: rpcclient/errors.go (hunk 7 above)
Comment (yyforyongyu) [reply rpcclient/errors.go]: TILed!

**Pull request merged**